File: integration-test/test/RegistrationSpec.js
'use strict'

const faker = require('faker')
const browser = require('../lib/BrowserBuilder')

const Url = {
  login: '/login',
  confirm: (token) => `/confirm/${token}`,
}

class RegistrationShortcuts {
  constructor(browser) {
    this.b = browser
  }

  async tryLogIn(email, password) {
    await this.b.assertExists('.session-form', { wait: 'pageLoad' })
    await this.b.sendKeys(email, '.session-form [name=email]')
    await this.b.sendKeys(password, '.session-form [name=password]')
    await this.b.click('.session-form [type=submit]')
  }

  async tryRegister(email, password, password2) {
    await this.b.assertExists('.user-form', { wait: 'pageLoad' })
    await this.b.sendKeys(email, '.user-form [name=email]')
    await this.b.sendKeys(password, '.user-form [name=password]')
    await this.b.sendKeys(password2, '.user-form [name=password2]')
    await this.b.click('.user-form [type=submit]')
  }

  async assertLoggedInAs(email) {
    await this.b.assertExists({ tag: 'div', class: 'logged-in', contains: email, wait: 'pageLoad' })
  }

  async assertLogInFailed() {
    await this.b.assertExists({ tag: 'h1', contains: 'Log in', wait: 'pageLoad' })
    await this.b.assertExists({ class: 'error', contains: 'Wrong email address or password' })
  }

  async logOut() {
    await this.b.click({ link: 'Log out' })
    await this.b.assertExists({ class: 'session-form', wait: 'pageLoad' })
  }
}

describe('Registration', function() {
  beforeEach(async function() {
    this.userEmail = faker.internet.email()
    this.validPassword = 'icrucGofbap4'

    this.browser = await browser.createBrowser()
    this.b = this.browser
    this.adminSession = browser.createUserAdminSession('Registration')
    this.registration = new RegistrationShortcuts(this.browser)
  })

  afterEach(async function() {
    if (this.browser) await this.browser.close()
  })

  it('should not register a bad email format', async function() {
    await this.b.get(Url.login)
    await this.registration.tryRegister('@' + this.userEmail, this.validPassword, this.validPassword)
    await this.b.assertExists('.user-form [name=email]:invalid')
  })

  it('should not register a weak password', async function() {
    await this.b.get(Url.login)
    await this.registration.tryRegister(this.userEmail, 'weak', 'weak')
    await this.b.assertExists('.user-form .control-group.error [name=password]')
    await this.b.assertExists('.user-form .control-group.error p.help-block')
  })

  it('should not register mismatched passwords', async function() {
    await this.b.get(Url.login)
    await this.registration.tryRegister(this.userEmail, this.validPassword, this.validPassword + '1')
    await this.b.assertExists('.user-form .control-group.error [name=password2]')
    await this.b.assertExists('.user-form .control-group.error p.help-block')
  })

  it('should show an error when clicking an invalid token link', async function() {
    await this.b.get(Url.confirm('abcdef'))
    await this.b.assertExists({ tag: 'h1', contains: 'Broken confirmation link' })
    await this.b.assertExists('a[href="/reset-password"]')
  })

  describe('when the user already exists', function() {
    beforeEach(async function() {
      await this.adminSession.createUser({
        email: this.userEmail,
        password: this.validPassword,
      })

      await this.b.get(Url.login)
      await this.registration.tryRegister(this.userEmail, this.validPassword + '1', this.validPassword + '1')
      await this.b.assertExists({ tag: 'h1', contains: 'Check your email', wait: 'pageLoad' })
    })

    afterEach(async function() {
      await this.adminSession.deleteUser({ email: this.userEmail })
    })

    it('should not change the password', async function() {
      // This would be a huge security hole!
      await this.b.get(Url.login)
      await this.registration.tryLogIn(this.userEmail, this.validPassword)
      await this.registration.assertLoggedInAs(this.userEmail)
      await this.registration.logOut()
      await this.registration.tryLogIn(this.userEmail, this.validPassword + '1')
      await this.registration.assertLogInFailed()
    })

    it('should not give the user a confirmation token', async function() {
      const json = await this.adminSession.showUser({ email: this.userEmail })
      const token = JSON.parse(json).confirmation_token

      expect(token).to.eq(null)
    })
  })

  describe('when the user tries to register', function() {
    beforeEach(/* beware -- not async */ function () {
      this.getConfirmationToken = async () => {
        const json = await this.adminSession.showUser({ email: this.userEmail })
        return JSON.parse(json).confirmation_token
      }
    })

    beforeEach(async function() {
      await this.b.get(Url.login)
      await this.registration.tryRegister(this.userEmail, this.validPassword, this.validPassword)
      await this.b.assertExists({ tag: 'h1', contains: 'Check your email', wait: 'pageLoad' })
    })

    afterEach(async function() {
      await this.adminSession.deleteUser({ email: this.userEmail })
    })

    it('should prompt the user to check his or her email', async function() {
      await this.b.assertExists({ tag: 'h1', contains: 'Check your email', wait: 'pageLoad' })
    })

    it('should set a confirmation token on the user', async function() {
      await this.getConfirmationToken().should.eventually.match(/\w+/)
    })

    it('should not let the user log in before confirming', async function() {
      await this.b.get('/login')
      await this.registration.tryLogIn(this.userEmail, this.validPassword)
      await this.b.assertExists({ class: 'error', contains: 'click the link in the confirmation email', wait: 'pageLoad' })
    })

    it('should recognize the user if the password is incorrect', async function() {
      await this.b.get('/login')
      await this.registration.tryLogIn(this.userEmail, this.validPassword + '1')
      await this.registration.assertLogInFailed()
    })

    describe('after clicking the confirmation link', function() {
      beforeEach(async function() {
        const token = await this.getConfirmationToken()
        await this.browser.get(Url.confirm(token))
      })

      it('should log you in and tell you all is well', async function() {
        await this.registration.assertLoggedInAs(this.userEmail)
        await this.b.assertExists({ class: 'alert-success', contains: 'Welcome' })
      })

      it('should disable the confirmation token', async function() {
        await this.getConfirmationToken().should.eventually.not.exist
      })
    })
  })
})

File: integration-test/spec/registration_spec.rb
#!/usr/bin/env ruby

require './spec/spec_helper'

# In order to let users store their own information
# We need to let users identify themselves
# By creating accounts
describe 'Registration' do
  TEST_USER_EMAIL = 'test-RANDOM@example.org'
  TEST_USER_PASSWORD = 'created during test'

  before do
    # start the admin session first. That way, when we start the user session
    # in run-browser, we'll see the user session atop the admin session. (In
    # most tests this is implicit, but since we don't create a user with the
    # admin session in this one, we have to be explicit here.)
    admin_session # will open the admin section

    # @user is not logged in (unlike other tests)
    @user = {
      email: TEST_USER_EMAIL.sub(/RANDOM/, rand(9999999999).to_s),
      password: TEST_USER_PASSWORD
    }
    page.visit('/login')
  end

  def assert_logged_in_as(email)
    page.assert_selector('div.logged-in', text: email, wait: WAIT_LOAD)
  end

  def assert_log_in_failed
    page.assert_selector('h1', text: /Log in/i, wait: WAIT_LOAD)
    page.assert_selector('.error', text: 'Wrong email address or password')
  end

  def try_log_in(email, password)
    page.within('.session-form') do
      page.fill_in('Email', with: email)
      page.fill_in('Password', with: password)
      page.click_button('Log in')
    end
  end

  def try_register(email, password, password2)
    page.within('.user-form') do
      page.fill_in('Email', with: email)
      page.fill_in('Password', with: password)
      page.fill_in('Password (again)', with: password2)
      page.click_button('Register and send confirmation email')
    end
  end

  def log_out
    page.click_link('Log out')
  end

  def get_confirmation_token(user)
    admin_session.get_user_token(user, 'confirmation')
  end

  it 'should not register a bad email format' do
    try_register(@user[:email] + '@foo', @user[:password], @user[:password])
    page.assert_selector('.user-form [name=email]:invalid')
  end

  it 'should not register a weak password' do
    try_register(@user[:email] + '@foo', 'weak', 'weak')
    page.assert_selector('p.help-block', text: 'characters long')
  end

  it 'should not register mismatched passwords' do
    try_register(@user[:email], @user[:password], @user[:password] + '1')
    page.assert_selector('p.help-block', text: 'doesn’t match')
  end

  it 'should error on invalid token link' do
    page.visit('/confirm/invalidtoken')
    page.assert_selector('h1', text: /Broken confirmation link/i, wait: WAIT_LOAD) # wait for page to load
  end

  describe 'when the user already exists' do
    before do
      @user = admin_session.create_test_user
      try_register(@user[:email], @user[:password] + '1', @user[:password] + '1')
    end

    after do
      admin_session.refresh
      admin_session.assert_selector('tr:not(.updating) td.email', text: @user[:email], wait: WAIT_LOAD) # wait for users to load
      admin_session.destroy_test_user(@user)
    end

    it 'should not indicate to an attacker that the account already exists' do
      page.assert_selector('h1', text: /Check your email/i, wait: WAIT_LOAD)
    end

    it 'should not let the attacker set a new password' do
      page.visit('/login')
      try_log_in(@user[:email], @user[:password] + '1')
      page.assert_selector('.help-block', text: 'Wrong email address or password')
    end

    it 'should not let the attacker prevent the good user from logging in' do
      page.visit('/login')
      try_log_in(@user[:email], @user[:password])
      assert_logged_in_as(@user[:email])
    end

    it 'should not create a confirmation token' do
      assert_equal(nil, get_confirmation_token(@user))
    end
  end

  describe 'when creating a new user' do
    before do
      try_register(@user[:email], @user[:password], @user[:password])
    end

    after do
      admin_session.refresh
      admin_session.assert_selector('tr:not(.updating) td.email', text: @user[:email], wait: WAIT_LOAD) # wait for users to load
      admin_session.destroy_test_user(@user)
    end

    it 'should prompt the user to check his or her email' do
      page.assert_selector('h1', text: /Check your email/i, wait: WAIT_LOAD) # wait for page to load
    end

    it 'should set a confirmation token on the user' do
      page.assert_selector('h1', text: /Check your email/i, wait: WAIT_LOAD) # wait for HTTP request to complete
      assert_match(/\A\w+\Z/, get_confirmation_token(@user))
    end

    it 'should not let the user log in yet' do
      page.visit('/login')
      try_log_in(@user[:email], @user[:password])
      page.assert_selector('h1', text: /Log in/i, wait: WAIT_LOAD)
      page.assert_selector('.error', text: 'click the link in the confirmation email')
    end

    describe 'after confirming' do
      before do
        page.assert_selector('h1', text: /Check your email/i, wait: WAIT_LOAD) # wait for HTTP request to complete
        page.visit("/confirm/#{get_confirmation_token(@user)}")
      end

      it 'should log you in and tell you all is well' do
        assert_logged_in_as(@user[:email])
        page.assert_selector('.alert-success', text: 'Welcome')
      end

      it 'should disable the confirmation token' do
        assert_equal(nil, get_confirmation_token(@user))
      end
    end
  end
end
